File: arrays&hashing/validanagram.py
# Input: s = "racecar", t = "carrace"

# Output: true


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        anagram_dict = {}
        for i in s:
            if i in anagram_dict:
                anagram_dict[i] = anagram_dict[i]+1
            else:
                anagram_dict[i]=1
        print(anagram_dict)


        for i in t:
            if i in anagram_dict:
                anagram_dict[i]=anagram_dict[i]-1
            else:
                return False
        print(anagram_dict)
        
        for k,v in anagram_dict.items():
            if v!=0:
                return False

        return True



sol =Solution()
s='racecar'
t='carrace'
print(sol.isAnagram(s,t))

File: leetcode/arrays&hashing/validanagram.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isAnagram(String s, String t) {
        Map<String, Integer> alphaCountMap= new HashMap<>();
        for(int i=0;i<s.length();i++)
        {   
            String chr  = Character.toString(s.charAt(i));
            if(alphaCountMap.containsKey(chr))
            {
                alphaCountMap.put(chr, alphaCountMap.get(chr)+1);
            }
            else
            {
                alphaCountMap.put(chr,1);
            }
        }

        for(int i=0;i<t.length();i++)
        {
            String chr  = Character.toString(t.charAt(i));
            if(alphaCountMap.containsKey(chr))
            {
                alphaCountMap.put(chr, alphaCountMap.get(chr)-1);
            }
            else
            {
                return false;
            }
        }

        for (String key : alphaCountMap.keySet())
        {
            if(alphaCountMap.get(key)>0)
            {
                return false;
            }
        }

        
        return true;
    }



    public static void main(String[] args)
    {
    String s="racecar";
    String t="carrace";
    Solution sol = new Solution();
    Boolean result = sol.isAnagram(s, t);
    System.out.println(result);
    }

}
